src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
         {
-            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenPrivateKeyFromEngine(TpmTssEngineName, TpmTssEngineEcDsaKeyHandle);
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenPrivateKeyFromEngine(TpmTssEngineName, TpmEcDsaKeyHandle);
             using ECDsa ecdsaPri = new ECDsaOpenSsl(priKeyHandle);
             using ECDsa ecdsaBad = ECDsa.Create();
             ecdsaBad.KeySize = ecdsaPri.KeySize;
 
             byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
             byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
             byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+            Assert.Equal(signature.Length, badSignature.Length);
             Assert.NotEqual(data, signature);
             Assert.True(ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
             Assert.False(ecdsaPri.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
         }
+
+        [ConditionalFact(nameof(ShouldRunProviderEcDsaTests))]
+        public static void Provider_TPM2ECDSA()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmEcDsaKeyHandleUri);
+            using ECDsa ecdsaPri = new ECDsaOpenSsl(priKeyHandle);
+
+            byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
+
+            byte[] ecdsaPubBytes = ecdsaPri.ExportSubjectPublicKeyInfo();
+            ECDsa ecdsaPub = ECDsa.Create();
+            ecdsaPub.ImportSubjectPublicKeyInfo(ecdsaPubBytes, out int bytesRead);
+            Assert.Equal(ecdsaPubBytes.Length, bytesRead);
+
+            using ECDsa ecdsaBad = ECDsa.Create();
+            ecdsaBad.KeySize = ecdsaPri.KeySize;
+
+            // Verify can sign/verify multiple times
+            for (int i = 0; i < 10; i++)
+            {
+                data[0] = (byte)i;
+                byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
+                byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+                Assert.NotEqual(data, signature);
+                Assert.NotEqual(data, badSignature);
+                Assert.NotEqual(badSignature, signature);
+                Assert.True(ecdsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaBad.VerifyData(data, signature, HashAlgorithmName.SHA256));
+
+                // TPM key is intended for sign/decrypt only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation but it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
```

Comment:
I'm of two minds here.  On the one hand, I can understand TPM2/OSSL's position "private keys are for signing, public keys are for verifying".

On the other hand, Verify works on literally every (non-Disposed) ECDsa instance until now... I believe that Windows makes this work in the NCrypt layer by just making a software public key when needed... so TPM keys work on Windows for both sign/verify... just not on OpenSSL.

That can be done as a followup, especially if it seems like something we want to debate.  (If deferring it, create an issue)

Reply:
My main problem with making that work is that it won't use the chosen provider anymore and it would just go through `default`. I'll open an issue to have that conversation separately.

Reply:
Opened: https://github.com/dotnet/runtime/issues/105174 to continue this conversation